32bit build fails on Linux i686

I was trying to build a package which depends on github.com/google/gopacket/pcap which depends on this package. I need (yes I know...) really need to run it on a 32bit Linux install but the build fails:

```
# github.com/mdlayher/raw
../../mdlayher/raw/raw_linux.go:316:32: undefined: syscall.SYS_GETSOCKOPT
../../mdlayher/raw/raw_linux.go:329:32: undefined: syscall.SYS_SETSOCKOPT
../../mdlayher/raw/timeval.go:17:6: cannot use int64(timeout / time.Second) (type int64) as type int32 in field value
../../mdlayher/raw/timeval.go:18:7: cannot use int64(timeout % time.Second / time.Microsecond) (type int64) as type int32 in field value
```

I am aware that a i386 build is not a priority in 2018 but this could probably be solved with just build costraints and splitting out the statistics into a separate file?

So instead of `// +build linux` use '// +build linux,!386` for 64bit and `// +build linux,386` for the 32bit files? Unfortunately I am a Go beginner so testing this enough to present a patch is proving a bit painful for me...

--- a/.travis.yml
+++ b/.travis.yml
@@ -5,7 +5,7 @@ os:
   - linux
   - osx
 before_install:
-  - go get github.com/golang/lint/golint
+  - go get golang.org/x/lint/golint
   - go get honnef.co/go/tools/cmd/staticcheck
   - go get -d -t ./...
 script:

--- a/raw_linux.go
+++ b/raw_linux.go
@@ -7,7 +7,6 @@ import (
 	"os"
 	"sync"
 	"sync/atomic"
-	"syscall"
 	"time"
 	"unsafe"
 
@@ -44,12 +43,12 @@ type packetConn struct {
 // socket is an interface which enables swapping out socket syscalls for
 // testing.
 type socket interface {
-	Bind(syscall.Sockaddr) error
+	Bind(unix.Sockaddr) error
 	Close() error
 	FD() int
 	GetSockopt(level, name int, v unsafe.Pointer, l uintptr) error
-	Recvfrom([]byte, int) (int, syscall.Sockaddr, error)
-	Sendto([]byte, int, syscall.Sockaddr) error
+	Recvfrom([]byte, int) (int, unix.Sockaddr, error)
+	Sendto([]byte, int, unix.Sockaddr) error
 	SetSockopt(level, name int, v unsafe.Pointer, l uint32) error
 	SetTimeout(time.Duration) error
 }
@@ -61,13 +60,13 @@ func listenPacket(ifi *net.Interface, proto uint16, cfg Config) (*packetConn, er
 	pbe := htons(proto)
 
 	// Enabling overriding the socket type via config.
-	typ := syscall.SOCK_RAW
+	typ := unix.SOCK_RAW
 	if cfg.LinuxSockDGRAM {
-		typ = syscall.SOCK_DGRAM
+		typ = unix.SOCK_DGRAM
 	}
 
 	// Open a packet socket using specified socket and protocol types.
-	sock, err := syscall.Socket(syscall.AF_PACKET, typ, int(pbe))
+	sock, err := unix.Socket(unix.AF_PACKET, typ, int(pbe))
 	if err != nil {
 		return nil, err
 	}
@@ -92,7 +91,7 @@ func newPacketConn(ifi *net.Interface, s socket, pbe uint16) (*packetConn, error
 	// packet(7):
 	//   Only the sll_protocol and the sll_ifindex address fields are used for
 	//   purposes of binding.
-	err := s.Bind(&syscall.SockaddrLinklayer{
+	err := s.Bind(&unix.SockaddrLinklayer{
 		Protocol: pbe,
 		Ifindex:  ifi.Index,
 	})
@@ -114,9 +113,9 @@ func (p *packetConn) ReadFrom(b []byte) (int, net.Addr, error) {
 	p.timeoutMu.Unlock()
 
 	var (
-		// Information returned by syscall.Recvfrom.
+		// Information returned by unix.Recvfrom.
 		n    int
-		addr syscall.Sockaddr
+		addr unix.Sockaddr
 		err  error
 
 		// Timeout for a single loop iteration.
@@ -144,7 +143,7 @@ func (p *packetConn) ReadFrom(b []byte) (int, net.Addr, error) {
 		switch err {
 		case nil:
 			// Got data, break this loop shortly.
-		case syscall.EAGAIN:
+		case unix.EAGAIN:
 			// Hit a timeout, keep looping.
 			continue
 		default:
@@ -157,9 +156,9 @@ func (p *packetConn) ReadFrom(b []byte) (int, net.Addr, error) {
 	}
 
 	// Retrieve hardware address and other information from addr.
-	sa, ok := addr.(*syscall.SockaddrLinklayer)
+	sa, ok := addr.(*unix.SockaddrLinklayer)
 	if !ok || sa.Halen < 6 {
-		return n, nil, syscall.EINVAL
+		return n, nil, unix.EINVAL
 	}
 
 	// Use length specified to convert byte array into a hardware address slice.
@@ -181,7 +180,7 @@ func (p *packetConn) WriteTo(b []byte, addr net.Addr) (int, error) {
 	// Ensure correct Addr type.
 	a, ok := addr.(*Addr)
 	if !ok || a.HardwareAddr == nil || len(a.HardwareAddr) < 6 {
-		return 0, syscall.EINVAL
+		return 0, unix.EINVAL
 	}
 
 	// Convert hardware address back to byte array form.
@@ -193,8 +192,8 @@ func (p *packetConn) WriteTo(b []byte, addr net.Addr) (int, error) {
 	//   When you send packets it is enough to specify sll_family, sll_addr,
 	//   sll_halen, sll_ifindex, and sll_protocol. The other fields should
 	//   be 0.
-	// In this case, sll_family is taken care of automatically by syscall.
-	err := p.s.Sendto(b, 0, &syscall.SockaddrLinklayer{
+	// In this case, sll_family is taken care of automatically by unix.
+	err := p.s.Sendto(b, 0, &unix.SockaddrLinklayer{
 		Ifindex:  p.ifi.Index,
 		Halen:    uint8(len(a.HardwareAddr)),
 		Addr:     baddr,
@@ -235,14 +234,14 @@ func (p *packetConn) SetWriteDeadline(t time.Time) error {
 
 // SetBPF attaches an assembled BPF program to a raw net.PacketConn.
 func (p *packetConn) SetBPF(filter []bpf.RawInstruction) error {
-	prog := syscall.SockFprog{
+	prog := unix.SockFprog{
 		Len:    uint16(len(filter)),
-		Filter: (*syscall.SockFilter)(unsafe.Pointer(&filter[0])),
+		Filter: (*unix.SockFilter)(unsafe.Pointer(&filter[0])),
 	}
 
 	err := p.s.SetSockopt(
-		syscall.SOL_SOCKET,
-		syscall.SO_ATTACH_FILTER,
+		unix.SOL_SOCKET,
+		unix.SO_ATTACH_FILTER,
 		unsafe.Pointer(&prog),
 		uint32(unsafe.Sizeof(prog)),
 	)
@@ -309,26 +308,26 @@ type sysSocket struct {
 
 // Method implementations simply invoke the syscall of the same name, but pass
 // the file descriptor stored in the sysSocket as the socket to use.
-func (s *sysSocket) Bind(sa syscall.Sockaddr) error { return syscall.Bind(s.fd, sa) }
-func (s *sysSocket) Close() error                   { return syscall.Close(s.fd) }
-func (s *sysSocket) FD() int                        { return s.fd }
+func (s *sysSocket) Bind(sa unix.Sockaddr) error { return unix.Bind(s.fd, sa) }
+func (s *sysSocket) Close() error                { return unix.Close(s.fd) }
+func (s *sysSocket) FD() int                     { return s.fd }
 func (s *sysSocket) GetSockopt(level, name int, v unsafe.Pointer, l uintptr) error {
-	_, _, err := syscall.Syscall6(syscall.SYS_GETSOCKOPT, uintptr(s.fd), uintptr(level), uintptr(name), uintptr(v), uintptr(unsafe.Pointer(&l)), 0)
+	_, _, err := unix.Syscall6(unix.SYS_GETSOCKOPT, uintptr(s.fd), uintptr(level), uintptr(name), uintptr(v), uintptr(unsafe.Pointer(&l)), 0)
 	if err != 0 {
-		return syscall.Errno(err)
+		return unix.Errno(err)
 	}
 	return nil
 }
-func (s *sysSocket) Recvfrom(p []byte, flags int) (int, syscall.Sockaddr, error) {
-	return syscall.Recvfrom(s.fd, p, flags)
+func (s *sysSocket) Recvfrom(p []byte, flags int) (int, unix.Sockaddr, error) {
+	return unix.Recvfrom(s.fd, p, flags)
 }
-func (s *sysSocket) Sendto(p []byte, flags int, to syscall.Sockaddr) error {
-	return syscall.Sendto(s.fd, p, flags, to)
+func (s *sysSocket) Sendto(p []byte, flags int, to unix.Sockaddr) error {
+	return unix.Sendto(s.fd, p, flags, to)
 }
 func (s *sysSocket) SetSockopt(level, name int, v unsafe.Pointer, l uint32) error {
-	_, _, err := syscall.Syscall6(syscall.SYS_SETSOCKOPT, uintptr(s.fd), uintptr(level), uintptr(name), uintptr(v), uintptr(l), 0)
+	_, _, err := unix.Syscall6(unix.SYS_SETSOCKOPT, uintptr(s.fd), uintptr(level), uintptr(name), uintptr(v), uintptr(l), 0)
 	if err != 0 {
-		return syscall.Errno(err)
+		return unix.Errno(err)
 	}
 	return nil
 }
@@ -337,5 +336,5 @@ func (s *sysSocket) SetTimeout(timeout time.Duration) error {
 	if err != nil {
 		return err
 	}
-	return syscall.SetsockoptTimeval(s.fd, syscall.SOL_SOCKET, syscall.SO_RCVTIMEO, tv)
+	return unix.SetsockoptTimeval(s.fd, unix.SOL_SOCKET, unix.SO_RCVTIMEO, tv)
 }

--- a/raw_linux_test.go
+++ b/raw_linux_test.go
@@ -5,7 +5,6 @@ package raw
 import (
 	"bytes"
 	"net"
-	"syscall"
 	"testing"
 	"time"
 	"unsafe"
@@ -18,11 +17,11 @@ import (
 // Test to ensure that socket is bound with correct sockaddr_ll information
 
 type bindSocket struct {
-	bind syscall.Sockaddr
+	bind unix.Sockaddr
 	noopSocket
 }
 
-func (s *bindSocket) Bind(sa syscall.Sockaddr) error {
+func (s *bindSocket) Bind(sa unix.Sockaddr) error {
 	s.bind = sa
 	return nil
 }
@@ -44,7 +43,7 @@ func Test_newPacketConnBind(t *testing.T) {
 		t.Fatal(err)
 	}
 
-	sall, ok := s.bind.(*syscall.SockaddrLinklayer)
+	sall, ok := s.bind.(*unix.SockaddrLinklayer)
 	if !ok {
 		t.Fatalf("bind sockaddr has incorrect type: %T", s.bind)
 	}
@@ -60,11 +59,11 @@ func Test_newPacketConnBind(t *testing.T) {
 // Test for incorrect sockaddr type after recvfrom on a socket.
 
 type addrRecvfromSocket struct {
-	addr syscall.Sockaddr
+	addr unix.Sockaddr
 	noopSocket
 }
 
-func (s *addrRecvfromSocket) Recvfrom(p []byte, flags int) (int, syscall.Sockaddr, error) {
+func (s *addrRecvfromSocket) Recvfrom(p []byte, flags int) (int, unix.Sockaddr, error) {
 	return 0, s.addr, nil
 }
 
@@ -72,7 +71,7 @@ func Test_packetConnReadFromRecvfromInvalidSockaddr(t *testing.T) {
 	p, err := newPacketConn(
 		&net.Interface{},
 		&addrRecvfromSocket{
-			addr: &syscall.SockaddrInet4{},
+			addr: &unix.SockaddrInet4{},
 		},
 		0,
 	)
@@ -81,7 +80,7 @@ func Test_packetConnReadFromRecvfromInvalidSockaddr(t *testing.T) {
 	}
 
 	_, _, err = p.ReadFrom(nil)
-	if want, got := syscall.EINVAL, err; want != got {
+	if want, got := unix.EINVAL, err; want != got {
 		t.Fatalf("unexpected error:\n- want: %v\n-  got: %v", want, got)
 	}
 }
@@ -92,7 +91,7 @@ func Test_packetConnReadFromRecvfromInvalidHardwareAddr(t *testing.T) {
 	p, err := newPacketConn(
 		&net.Interface{},
 		&addrRecvfromSocket{
-			addr: &syscall.SockaddrLinklayer{
+			addr: &unix.SockaddrLinklayer{
 				Halen: 5,
 			},
 		},
@@ -103,7 +102,7 @@ func Test_packetConnReadFromRecvfromInvalidHardwareAddr(t *testing.T) {
 	}
 
 	_, _, err = p.ReadFrom(nil)
-	if want, got := syscall.EINVAL, err; want != got {
+	if want, got := unix.EINVAL, err; want != got {
 		t.Fatalf("unexpected error:\n- want: %v\n-  got: %v", want, got)
 	}
 }
@@ -113,11 +112,11 @@ func Test_packetConnReadFromRecvfromInvalidHardwareAddr(t *testing.T) {
 type recvfromSocket struct {
 	p     []byte
 	flags int
-	addr  syscall.Sockaddr
+	addr  unix.Sockaddr
 	noopSocket
 }
 
-func (s *recvfromSocket) Recvfrom(p []byte, flags int) (int, syscall.Sockaddr, error) {
+func (s *recvfromSocket) Recvfrom(p []byte, flags int) (int, unix.Sockaddr, error) {
 	copy(p, s.p)
 	s.flags = flags
 	return len(s.p), s.addr, nil
@@ -130,7 +129,7 @@ func Test_packetConnReadFromRecvfromOK(t *testing.T) {
 
 	s := &recvfromSocket{
 		p: data,
-		addr: &syscall.SockaddrLinklayer{
+		addr: &unix.SockaddrLinklayer{
 			Halen: 6,
 			Addr:  [8]byte{0xde, 0xad, 0xbe, 0xef, 0xde, 0xad, 0x00, 0x00},
 		},
@@ -176,7 +175,7 @@ func Test_packetConnReadFromRecvfromOK(t *testing.T) {
 
 func Test_packetConnWriteToInvalidSockaddr(t *testing.T) {
 	_, err := (&packetConn{}).WriteTo(nil, &net.IPAddr{})
-	if want, got := syscall.EINVAL, err; want != got {
+	if want, got := unix.EINVAL, err; want != got {
 		t.Fatalf("unexpected error:\n- want: %v\n-  got: %v", want, got)
 	}
 }
@@ -195,7 +194,7 @@ func Test_packetConnWriteToInvalidHardwareAddr(t *testing.T) {
 		_, err := (&packetConn{}).WriteTo(nil, &Addr{
 			HardwareAddr: addr,
 		})
-		if want, got := syscall.EINVAL, err; want != got {
+		if want, got := unix.EINVAL, err; want != got {
 			t.Fatalf("unexpected error:\n- want: %v\n-  got: %v", want, got)
 		}
 	}
@@ -206,11 +205,11 @@ func Test_packetConnWriteToInvalidHardwareAddr(t *testing.T) {
 type sendtoSocket struct {
 	p     []byte
 	flags int
-	addr  syscall.Sockaddr
+	addr  unix.Sockaddr
 	noopSocket
 }
 
-func (s *sendtoSocket) Sendto(p []byte, flags int, to syscall.Sockaddr) error {
+func (s *sendtoSocket) Sendto(p []byte, flags int, to unix.Sockaddr) error {
 	copy(s.p, p)
 	s.flags = flags
 	s.addr = to
@@ -254,7 +253,7 @@ func Test_packetConnWriteToSendtoOK(t *testing.T) {
 		t.Fatalf("unexpected data:\n- want: %v\n-  got: %v", want, got)
 	}
 
-	sall, ok := s.addr.(*syscall.SockaddrLinklayer)
+	sall, ok := s.addr.(*unix.SockaddrLinklayer)
 	if !ok {
 		t.Fatalf("write sockaddr has incorrect type: %T", s.addr)
 	}
@@ -330,10 +329,10 @@ func Test_packetConnSetBPF(t *testing.T) {
 	fn := func(level, name int, _ unsafe.Pointer, _ uint32) error {
 		// Though we can't check the filter itself, we can check the setsockopt
 		// level and name for correctness.
-		if want, got := syscall.SOL_SOCKET, level; want != got {
+		if want, got := unix.SOL_SOCKET, level; want != got {
 			t.Fatalf("unexpected setsockopt level:\n- want: %v\n-  got: %v", want, got)
 		}
-		if want, got := syscall.SO_ATTACH_FILTER, name; want != got {
+		if want, got := unix.SO_ATTACH_FILTER, name; want != got {
 			t.Fatalf("unexpected setsockopt name:\n- want: %v\n-  got: %v", want, got)
 		}
 
@@ -359,8 +358,8 @@ type setTimeoutSocket struct {
 	noopSocket
 }
 
-func (s *setTimeoutSocket) Recvfrom(p []byte, flags int) (int, syscall.Sockaddr, error) {
-	return 0, &syscall.SockaddrLinklayer{
+func (s *setTimeoutSocket) Recvfrom(p []byte, flags int) (int, unix.Sockaddr, error) {
+	return 0, &unix.SockaddrLinklayer{
 		Halen: 6,
 	}, nil
 }
@@ -445,11 +444,11 @@ func Test_packetConn_handleStats(t *testing.T) {
 // the basis for more specific socket implementations.
 type noopSocket struct{}
 
-func (noopSocket) Bind(sa syscall.Sockaddr) error                                { return nil }
+func (noopSocket) Bind(sa unix.Sockaddr) error                                   { return nil }
 func (noopSocket) Close() error                                                  { return nil }
 func (noopSocket) FD() int                                                       { return 0 }
 func (noopSocket) GetSockopt(level, name int, v unsafe.Pointer, l uintptr) error { return nil }
-func (noopSocket) Recvfrom(p []byte, flags int) (int, syscall.Sockaddr, error)   { return 0, nil, nil }
-func (noopSocket) Sendto(p []byte, flags int, to syscall.Sockaddr) error         { return nil }
+func (noopSocket) Recvfrom(p []byte, flags int) (int, unix.Sockaddr, error)      { return 0, nil, nil }
+func (noopSocket) Sendto(p []byte, flags int, to unix.Sockaddr) error            { return nil }
 func (noopSocket) SetSockopt(level, name int, v unsafe.Pointer, l uint32) error  { return nil }
 func (noopSocket) SetTimeout(timeout time.Duration) error                        { return nil }

--- a/timeval.go
+++ b/timeval.go
@@ -1,4 +1,4 @@
-// +build !darwin,!arm,!windows,!mipsle,!mips
+// +build !darwin,!arm,!windows,!mipsle,!mips,!386
 
 package raw
 

--- a/timeval.go
+++ b/timeval.go
@@ -3,17 +3,18 @@
 package raw
 
 import (
-	"syscall"
 	"time"
+
+	"golang.org/x/sys/unix"
 )
 
-// newTimeval transforms a duration into a syscall.Timeval struct.
+// newTimeval transforms a duration into a unix.Timeval struct.
 // An error is returned in case of zero time value.
-func newTimeval(timeout time.Duration) (*syscall.Timeval, error) {
+func newTimeval(timeout time.Duration) (*unix.Timeval, error) {
 	if timeout < time.Microsecond {
 		return nil, &timeoutError{}
 	}
-	return &syscall.Timeval{
+	return &unix.Timeval{
 		Sec:  int64(timeout / time.Second),
 		Usec: int64(timeout % time.Second / time.Microsecond),
 	}, nil

--- a/timeval32.go
+++ b/timeval32.go
@@ -1,19 +1,20 @@
-// +build arm mipsle mips
+// +build arm mipsle mips 386
 
 package raw
 
 import (
-	"syscall"
 	"time"
+
+	"golang.org/x/sys/unix"
 )
 
-// newTimeval transforms a duration into a syscall.Timeval struct.
+// newTimeval transforms a duration into a unix.Timeval struct.
 // An error is returned in case of zero time value.
-func newTimeval(timeout time.Duration) (*syscall.Timeval, error) {
+func newTimeval(timeout time.Duration) (*unix.Timeval, error) {
 	if timeout < time.Microsecond {
 		return nil, &timeoutError{}
 	}
-	return &syscall.Timeval{
+	return &unix.Timeval{
 		Sec:  int32(timeout / time.Second),
 		Usec: int32(timeout % time.Second / time.Microsecond),
 	}, nil

